### PR TITLE
Overhaul ju.Date: always use the GMT time zone, and implement it without JS interop.

### PR DESCRIPTION
**Pull request by sjrd**
This is a policy decision, which standardizes on GMT as the default timezone for every Scala.js codebase.

This policy is in line with other things thar are normally environment-dependent on the JVM:

* The default locale is always `ROOT`.
* The default encoding is always UTF-8.
* The default line separator is always `\n`.
* And now the default timezone is always Etc/GMT.

---

The implementation without JS interop also fixes several issues that are unfixable with `js.Date`:

* We are freed from the range limitation of `js.Date`.
* We correctly handle the Julian half of the timeline (`js.Date` is Gregorian-only, while `ju.Date` is Julian/Gregorian).
* We do not suffer from subtle precision errors surfaced by the (specified) calculations in `Double`s of `js.Date`.

In addition, this implementation will work in Wasm without a JS host.

---

~~WiP because needs more tests, especially of the Julian calendar and of `set*` methods.~~

### CODE REVIEW

**Comment by sjrd** [comment]
@gzm0 Not ready for review yet, but WDYT of the policy change to standardize on Etc/GMT (i.e., the canonical administrative timezone that sticks to UTC without daylight saving time)?

**Comment by gzm0** [comment]
Hrm... The Javadoc for Date mentions the [local time zone](https://docs.oracle.com/en/java/javase/26/docs/api/java.base/java/util/Date.html#:~:text=min%20arguments%2C%20in%20the-,local%20time%20zone,-.), not the [default time zone](https://docs.oracle.com/en/java/javase/26/docs/api/java.base/java/util/TimeZone.html#getDefault()) (it is not clear to me whether they are equivalent).

Are you sure it is worth investing time into this API? It feels it might be better to just not make it link on WASM?

**Comment by sjrd** [comment]
> Hrm... The Javadoc for Date mentions the [local time zone](https://docs.oracle.com/en/java/javase/26/docs/api/java.base/java/util/Date.html#:~:text=min%20arguments%2C%20in%20the-,local%20time%20zone,-.), not the [default time zone](https://docs.oracle.com/en/java/javase/26/docs/api/java.base/java/util/TimeZone.html#getDefault()) (it is not clear to me whether they are equivalent).

Experimentally, setting `user.timezone`, which explicitly controls the default timezone, also affects `ju.Date`. So yes, they must be the same.

> Are you sure it is worth investing time into this API? It feels it might be better to just not make it link on WASM?

Maybe not. If we don't, it will forever stick like a store thumb, though. Because it would be the *only* part of the javalib that we would never support on Wasm.

**Comment by gzm0** [comment]
> Because it would be the only part of the javalib that we would never support on Wasm.

Ah right, we don't want that :P In this case, given this only affects deprecated APIs; I think specifying GMT is fine.

**Comment by sjrd** [comment]
Some more reasons to do this: note that `ju.Date` per se is not deprecated. It is also used as input/output type for other APIs that have no deprecation or replacement. Only the calendar-sensitive methods are deprecated. But even then, `toString()` isn't (and cannot, obviously), and it must link. So we cannot avoid implementing the decoding aspects of the class in order to implement `toString()` from a timestamp.

We could get away with not linking `parse()` on Wasm. But given the current state even for JS, I believe it's worth a reimplementation.

---

Also: this is such a good excuse to exercise our `Long` divisions by constants :sweat_smile: There are cases of $0 < b < 2^{18}$ and cases of $2^{18} < b < 2^{31}$.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala [hunk 1]
```diff
@@ -84,11 +85,101 @@ class DateTest {
...
+    // Other separators
+    test(878707407000L, "5\tNov\n1997 5:23:27 GMT")
+    test(878707407000L, "1997, 5\rNov 5:23:27 GMT")
+    test(878707407000L, "5\tNov\n1997 5:23,27 GMT")
+
+    // Various prefixes
+    test(878707407000L, "No 5 1997 5:23:27 GMT")
+    test(876029007000L, "October 5 1997 sunday 5:23:27 GMT")
+    test(876029007000L, "Oct5 1997 sunday5:23:27 GMT")
+
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala:216]: Add some tests with duplicate non-digit / non-word characters (`,+-:/`)? At least IIUC that should be allowed.

File: javalib/src/main/scala/java/util/Date.scala [hunk 2]
```diff
@@ -191,8 +218,231 @@ object Date {
...
+        }
+        val number = Integer.parseInt(string.substring(start, i))
+
+        // Look behind and ahead
+        val prev = if (start == 0) EOF else string.charAt(start - 1).toInt
+        val next = if (i == len) EOF else string.charAt(i).toInt
+
+        // > The number is regarded as a year number if one of the following conditions is true:
+        @inline def isYear: Boolean = {
+          (year == -1) && // out of spec, but otherwise nothing works
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala]: I don't fully understand why nothing works without this condition. Maybe add an example?

I do wonder if the bullets in the JDK docs are to be interpreted jointly (not in order).

File: javalib/src/main/scala/java/util/Date.scala [hunk 3]
```diff
@@ -191,8 +218,231 @@ object Date {
...
+            }
+          }
+        }
+
+        // Decide what to do with the number
+        if ((prev == '+' || prev == '-') && year != -1) {
+          /* > If a number is preceded by + or - and a year has already been
+           * > recognized, then the number is a time-zone offset.
+           */
+          if (timezoneOffset != Int.MinValue && timezoneOffset != 0)
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala]: The second condition seems to be more permissive then what the spec says (provided we implicitly interpret that any given value may not be specified twice).

IIUC, this allows `+0 +1`, but the spec mentions words that are interpreted as time zones.

Does this match observed JDK behavior? Otherwise it would feel more natural to have a `timezone` and a `timezoneOffset` variable that only get assembled at the end:

> If a time zone or time-zone offset has been recognized, ...

File: javalib/src/main/scala/java/util/Date.scala [hunk 4]
```diff
@@ -191,8 +218,231 @@ object Date {
...
+      } else if (isASCIILetter(c)) {
+        // Parse the whole ASCII word
+        while (i != len && isASCIILetter(string.charAt(i))) {
+          i += 1
+        }
+        if (i == start + 1)
+          fail() // out of spec, but apparently the JVM wants at least 2 letters here
+        val word = string.substring(start, i).toUpperCase()
+
+        // Decide what to do with it
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala:747]: Looks like AM / PM handling is missing here? (also in the tests, obviously).

File: javalib/src/main/scala/java/util/Date.scala [hunk 5]
```diff
@@ -191,8 +218,231 @@ object Date {
...
+
+        val prefixWords = ParsePrefixWords // local copy
+        val prefixWordsLen = prefixWords.length
+        var prefixIndex = 0
+        while (prefixIndex != prefixWordsLen && !prefixWords(prefixIndex).startsWith(word)) {
+          prefixIndex += 1
+        }
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala:754]: How much worse (in terms of resulting code) is this to using `indexWhere`?

File: javalib/src/main/scala/java/util/Date.scala [hunk 6]
```diff
@@ -191,8 +218,231 @@ object Date {
...
+            // Ignore all characters until a matching ')'
+            var nesting = 1
+            while (i != len && nesting > 0) {
+              string.charAt(i) match {
+                case '(' => nesting += 1
+                case ')' => nesting -= 1
+                case _   => ()
+              }
+              i += 1
+            }
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala:799]: Fail if the parenthesis are not properly nested? (if the parenthesis are not nested, the material may not be ignored, leading to failure, because `(` is not a permitted ASCII character).

File: javalib/src/main/scala/java/util/Date.scala [hunk 7]
```diff
@@ -191,8 +218,231 @@ object Date {
...
+    if (year == -1 || month == -1 || dayOfMonth == -1)
+      fail()
+
+    // However, time components are optional and default to 0
+    if (hour == -1)
+      hour = 0
+    if (minute == -1)
+      minute = 0
+    if (second == -1)
+      second = 0
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala:821]: IIUC it cannot happen (based on the parsing rules) that more significant parts of the time are not set while less significant parts are. Consider commenting and/or asserting.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala [hunk 8]
```diff
@@ -84,11 +85,101 @@ class DateTest {
...
+    /* If there are several time zones, the last one wins
+     * (which is why GMT-0115 works without dedicated code paths, btw).
+     */
+    test(878711907000L, "Nov 5 1997 GMT 5:23:27 -0115")
+    test(878725407000L, "Nov 5 1997 +0 GMT -0115 EST 5:23:27")
+    test(878725407000L, "Nov 5 1997 +0 GMT EST 5:23:27")
+    test(878707407000L, "Nov 5 1997 +0 -0115 EST GMT 5:23:27")
+    test(878689407000L, "Nov 5 1997 +0 EST GMT 5:23:27 +5")
+
+    // But a time zone offset is rejected if the last time zone so far is non-UTC (not specified)
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala]: > It is permitted for the string to specify GMT, UT, or UTC redundantly-for example, GMT-5 or utc+0430.

Not sure I would call this "not specified". But clearly, the specification is unclear / inconsistent with the implementation here.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala [hunk 9]
```diff
@@ -84,11 +85,101 @@ class DateTest {
...
+    testFailure("Nov 5 1997 +0 -0115 +5 5:23:27")
+    testFailure("Nov 5 1997 +0 EST -0115 GMT 5:23:27")
+
+    // Some other formats
+    test(878707407000L, "5 Nov 1997 5:23:27 GMT")
+    test(878707407000L, "11/5/1997 5:23:27 GMT")
+    test(878707407000L, "Nov 5 Thu 1997 5:  23:  27 GMT") // It was a Wednesday, actually
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala]: Test whitespaces before relevant suffixes (e.g. `5 :3 :2 GMT` which I presume should fail).

File: javalib/src/main/scala/java/util/Date.scala [hunk 10]
```diff
@@ -191,8 +218,231 @@ object Date {
...
+  def parse(string: String): Long = {
+    /* Parse the string according to
+     * https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/util/Date.html#parse(java.lang.String)
+     * Comments starting with '>' are quoted from the JavaDoc.
+     */
+
+    /* We can use -1 as absent value for these fields because they cannot
+     * be parsed as negative values.
+     */
+    var year = -1
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala]: Nit: Comment that this is the full year, not the `-1900` adjusted value?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala:1]: FWIW: This might be a good candidate to use your favorite LLM to look for corner cases that are not covered. Lest IP pollution is a concern, of course.

File: javalib/src/main/scala/java/util/Date.scala [hunk 11]
```diff
@@ -214,7 +219,7 @@ object Date {
   @Deprecated
   def UTC(year: Int, month: Int, date: Int,
       hrs: Int, min: Int, sec: Int): Long = {
-    js.Date.UTC(year + 1900, month, date, hrs, min, sec).toLong
+    safeGetTime(js.Date.UTC(year + 1900, month, date, hrs, min, sec))
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala]: IIUC this is a (somewhat) unrelated bugfix? Put into a separate commit (together with the test change)?

**Review by gzm0 — CHANGES_REQUESTED** [review]
Review except for the last commit.

File: javalib/src/main/scala/java/util/Date.scala [hunk 12]
```diff
@@ -46,131 +43,312 @@ class Date(private var millis: Long)
...
+  override def hashCode(): Int =
+    java.lang.Long.hashCode(time)
+
+  // API that requires decoding and encoding into calendar fields
+
+  @inline
+  private def decode(): Decoded = {
+    if (cachedRest == 0)
+      doDecodeAndCache()
+    Decoded.fromBits(cachedYear, cachedRest)
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala:92]: This is a neat trick with the inline class :)

File: javalib/src/main/scala/java/util/Date.scala [hunk 13]
```diff
@@ -46,131 +43,312 @@ class Date(private var millis: Long)
...
+    val (year, dayInYear, isLeapYear) =
+      if (isGregorian) decodeGregorianYearAndDayInYear(daysSinceEpoch)
+      else decodeJulianYearAndDayInYear(daysSinceEpoch)
+
+    // Compute the month-in-year and day-in-month
+    val (month, day) = if (dayInYear < 31 + 28 + isLeapYear) {
+      if (dayInYear < 31)
+        (0, dayInYear)
+      else
+        (1, dayInYear - 31)
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala]: Unnest the if for readability?

```scala
if (dayInYear < 31) {
  (0, dayInYear)
} else if (dayInYear < 31 + 28 + isLeapYear) {
  (1, dayInYear - 31)
} else {
```

File: javalib/src/main/scala/java/util/Date.scala [hunk 14]
```diff
@@ -46,131 +43,312 @@ class Date(private var millis: Long)
...
   @Deprecated
   def getTimezoneOffset(): Int = 0
 
-  @Deprecated
-  def getYear(): Int = asDate().getUTCFullYear().toInt - 1900
+  /** Resets the year and month.
+   *
+   *  This requires a full re-encoding.
+   */
+  private def setYearMonth(d: Decoded, newYear: Int, newMonth: Int): Unit = {
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala:246]: It looks like you're passing a `Decoded` across an unspecified inline boundary. Intended?

File: javalib/src/main/scala/java/util/Date.scala [hunk 15]
```diff
@@ -200,7 +378,106 @@ object Date {
    *  lifetime of the application.
    */
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala:385]: Add a comment (probably in the introducing commit), why we make this a lazy val. (IIUC the optimizer cannot fully DCE this, because it cannot prove side-effect freedom).

File: javalib/src/main/scala/java/util/Date.scala [hunk 16]
```diff
@@ -216,10 +493,80 @@ object Date {
...
+    encode(year, month, day, timeInDay)
+  }
+
+  private def encode(year: Int, month: Int, day: Int, timeInDay: Long): Long = {
+    val daysSinceEpoch = encodeDaysSinceEpoch(year, month, day)
+    daysSinceEpoch * MSPerDay + timeInDay
+  }
+
+  @inline
+  private def encodeTimeInDay(hour: Int, min: Int, sec: Int, ms: Int): Long =
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala]: It looks like we never use `ms`? If that's indeed the case, remove?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala [hunk 17]
```diff
@@ -30,6 +30,16 @@ class DateTest {
...
+    // First Gregorian second and last Julian second
+    assertEquals(-12219292800000L, Date.UTC(1582 - 1900, 9, 15, 0, 0, 0))
+    assertEquals(-12219292801000L, Date.UTC(1582 - 1900, 9, 4, 23, 59, 59))
+
+    // The last second interpreted as Julian (but wrapping into the Gregorian calendar)
+    assertEquals(-12218428801000L, Date.UTC(1582 - 1900, 9, 14, 23, 59, 59))
+
+    assertEquals(94668649689600000L, Date.UTC(3000000, 1, 1, 0, 0, 0))
+    assertEquals(-94675005273600000L, Date.UTC(-3000000, 1, 1, 0, 0, 0))
   }
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala:56]: Add some tests with out of range values? (e.g. day 32, etc.)

**Review by gzm0 — CHANGES_REQUESTED** [review]
Complete review.

File: javalib/src/main/scala/java/util/Date.scala (hunk 14 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Date.scala:246]: Added `@inline`. That's clearer (although does not affect the generated code, because `Decoded`, being a virtualized instance, causes `setYearMonth` to be inlined anyway).

File: javalib/src/main/scala/java/util/Date.scala (hunk 2 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Date.scala]: Added some examples (which are in the tests).

> I do wonder if the bullets in the JDK docs are to be interpreted jointly (not in order).

I'm not sure what that would mean either. 🤔

File: javalib/src/main/scala/java/util/Date.scala (hunk 4 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Date.scala:747]: Ouch, good catch. Added.

File: javalib/src/main/scala/java/util/Date.scala (hunk 5 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Date.scala:754]: We don't have `indexWhere` on `Array`s in the javalib. At least not so far.

File: javalib/src/main/scala/java/util/Date.scala (hunk 3 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Date.scala]: > Does this match observed JDK behavior?

Yes, it does. I added one more test `"Nov 5 1997 +0 +5 5:23:27"` for that behavior.

File: javalib/src/main/scala/java/util/Date.scala (hunk 6 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Date.scala:799]: Tests added. There's a failure for too many closing parentheses (because `)` is not permitted). But if an open parenthesis is never closed, then everything is ignored until EOF.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala (hunk 8 above)
Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala]: I changed the comment from "not specified" to "out of spec". Is that clearer?

**Review by sjrd — COMMENTED** [review]
All right. I *think* I addressed all the comments.

I found a different way to deal with `year`, which probably makes more sense.

File: javalib/src/main/scala/java/util/Date.scala [hunk 18]
```diff
@@ -191,7 +242,245 @@ object Date {
...
+               */
+              if (number >= 70 && (next == ' ' || next == ',' || next == EOF))
+                interpretAsYear()
+              else if (hour != -1 && minute == -1)
+                minute = number
+              else if (minute != -1 && second == -1)
+                second = number
+              else if (dayOfMonth == -1)
+                dayOfMonth = number
+              else if (number < 70)
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala]: Shouldn't this be guarded by a month already having been recognized?

> - The number is less than 70, and both a month and a day of the
>   month have already been recognized.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala [hunk 19]
```diff
@@ -19,99 +19,546 @@ import org.junit.Assume._
...
-  }
+    // Unbalanced parentheses
+    test(878688000000L, "Nov 5 1997 GMT (ig(noré) 5:23:27") // not closed -> ignore until EOF
+    testFailure("Nov 5 1997 GMT (ig(no)ré)) 5:23:27") // too many closed -> failure
+
+    // Explicit time zone offsets
+    test(878689407000L, "Nov 5 1997 5:23:27 +5")
+    test(878414400000L, "Nov 1 1997 utc-20")
+    test(-11939000L, "Jan 1 1970 18:11:01 +2130")
+    test(878711907000L, "Nov 5 1997 (ig(no)ré) 5:23:27 GMT-0115")
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala:138]: We are missing tests where parenthesis are between a number and the following character. For example: `Nov 5 1997 5(ignored):23:27`.

File: javalib/src/main/scala/java/util/Date.scala [hunk 20]
```diff
@@ -191,7 +242,245 @@ object Date {
...
+                interpretAsYear()
+              else
+                fail()
+
+            case _ =>
+              // Otherwise fail (not clearly specified)
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Date.scala:734]: It seems we have insufficient test coverage for this:
- https://github.com/scala-js/scala-js/pull/5341/changes#r3608174894
- https://github.com/scala-js/scala-js/pull/5341/changes#r3608178366

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala [hunk 21]
```diff
@@ -19,99 +19,546 @@ import org.junit.Assume._
...
+    testFailure("11/5/7/ 6:23:27 GMT") // 7 is interpreted as dayOfMonth
 
-  @Test def before(): Unit = {
-    assertTrue(new Date(97, 11, 5, 0, 0).before(new Date(98, 11, 5, 0, 0)))
-    assertFalse(new Date(99, 11, 5, 0, 0).before(new Date(98, 11, 5, 0, 0)))
-    assertFalse(new Date(99, 11, 5, 0, 0).before(new Date(99, 11, 5, 0, 0)))
+    // Other separators
+    test(878707407000L, "5\tNov\n1997 5:23:27 GMT")
+    test(878707407000L, "1997, 5\rNov 5:23:27 GMT")
+    test(878707407000L, "5\tNov\n1997 5:23,27 GMT")
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DateTest.scala:195]: It seems we are missing tests where a number is followed by a permitted ascii letter (e.g. `5nov`).

**Review by gzm0 — CHANGES_REQUESTED** [review]
Nice! I think the year handling indeed is nicer this way.

I think I found some more special cases for `parse` :-/

File: javalib/src/main/scala/java/util/Date.scala (hunk 18 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Date.scala]: Good catch. Added a test as well.

**Review by sjrd — COMMENTED** [review]
Here come more tests :p

**Review by gzm0 — APPROVED** [review]
